### PR TITLE
Update SlotFillProvider to the new context API

### PR DESCRIPTION
**Pull request by oandregal**
To be merged after https://github.com/WordPress/gutenberg/pull/11314 lands.

The SlotFillProvider is using the React's [legacy context API](https://reactjs.org/docs/legacy-context.html) which will be removed in the next major version and was added as a [deprecation in StrictMode](https://reactjs.org/blog/2018/10/23/react-v-16-6.html#deprecations-in-strictmode) in 16.6.0. 

This PR updates the SlotFillProvider to the stable React's context API. 

## Test

- The popover component works as expected. For example, use the `/` inserter to add blocks. 
- Install a plugin that uses the SlotFill API (for example: DropIt or Yoast SEO) and make sure it works as expected.

### CODE REVIEW

File: packages/components/src/slot-fill/index.js
Comment (nosolosw) [inline packages/components/src/slot-fill/index.js]: I originally used `SlotFillConsumer` instead of `Consumer`. Changed the name as to be more in line with `Provider` naming.

File: packages/components/src/slot-fill/provider.js
Comment (nosolosw) [reply packages/components/src/slot-fill/provider.js]: No, they don't. This needs to be an object lifted up to the constructor, to prevent re-render consumers every time this component renders. I don't have any preference on use `state` on a different object.

**Comment by nosolosw** [comment]
> I'm thinking we should upgrade to React v16.6 first so that we can use Class.contextType to migrate these components to the new Context API without having to introduce new wrapper components.

I'd rather have this upgrade in 5.0. Using legacy APIs inside Gutenberg justifies its use within third-party (plugins, etc), which are a lot harder to migrate. Also, when we upgrade React we might end doing things differently ([hooks](https://reactjs.org/docs/hooks-intro.html) anyone?) where `Class.contextType` can't be used, or perhaps there is an even newer/better API built on top of the low-level primitives.

**Comment by youknowriad** [comment]
Moving out of 4.2 as it doesn't really touch the public API but still good to have for 5.0 among other react deprecations/upgrade.

**Comment by nosolosw** [comment]
Wanted to note that this PR is blocked until the tests are fixed.

The reason why they fail is that we are using an old version of enzyme that doesn't support the new React Context API. Updating `enzyme` and `enzyme-adapter-react-16` fixes the issue. I'll put together a PR to update those dependencies so we can explore their impact. Note that [enzyme not fully supporting React 16](https://github.com/airbnb/enzyme/issues/1553) has been a problem in the past that forced us to either [skip tests](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/block-controls/test/index.js#L30), [remove](https://github.com/WordPress/gutenberg/pull/6732), or [migrate](https://github.com/WordPress/gutenberg/commit/22ae6093899ae6ab2d99fa395206695499128618) them to react-test-renderer.

**Comment by nosolosw** [comment]
Update: after some slack convos, I've learned that we'd prefer to update those tests to the react-test-renderer, so I'm going to do so. 

Also, I've created https://github.com/WordPress/gutenberg/issues/11214 to keep track of workarounds we do to make enzyme and enzyme-adapter-react-16 work with React.

File: packages/components/src/slot-fill/test/__snapshots__/slot.js.snap
Comment (nosolosw) [inline packages/components/src/slot-fill/test/__snapshots__/slot.js.snap]: Note that, previously, we didn't do snapshot testing and the `button` component wasn't checked for in this test.

File: packages/components/src/slot-fill/test/__snapshots__/slot.js.snap
Comment (nosolosw) [inline packages/components/src/slot-fill/test/__snapshots__/slot.js.snap]: Note that, previously, we weren't doing snapshot testing and the `button` component wasn't checked for.

File: packages/components/src/slot-fill/test/__snapshots__/slot.js.snap
Comment (nosolosw) [inline packages/components/src/slot-fill/test/__snapshots__/slot.js.snap]: Note that, previously, we weren't doing snapshot testing and the `button` component wasn't checked for in this test.

**Comment by nosolosw** [comment]
Tests updated - note that tests are migrated in atomic commits one by one so it's easier to compare the changes. This is ready for a second review.

**Comment by nosolosw** [comment]
mmm, some other tests still need to be migrated.

**Comment by nosolosw** [comment]
Tests have been fixed. This is ready for review.

File: packages/components/src/slot-fill/provider.js
Comment (gziolo) [reply packages/components/src/slot-fill/provider.js]: It doesn't matter that much. It needs to be passed here by reference to avoid re-renders as @nosolosw pointed already.

File: packages/components/src/slot-fill/index.js
Comment (gziolo) [inline packages/components/src/slot-fill/index.js]: Should we rename it to context and export both components the same way? This could be also simplified to:
```js
export { SlotFillProvider as Provider, Consumer } from './context';
```

File: packages/components/src/slot-fill/slot.js
Comment (gziolo) [inline packages/components/src/slot-fill/slot.js]: It seems like `Slot` should be at the end of the file

File: packages/components/src/slot-fill/test/slot.js
Comment (gziolo) [inline packages/components/src/slot-fill/test/slot.js]: All those tests are much harder to read since snapshots were introduced. Is it possible to rewrite them to use inline check against generated HTML?

**Review by gziolo — CHANGES_REQUESTED** [review]
It would be nice to move all test updates to their own PR. Ensure it works with `master` and then upgrade Slot and Fill to the new Context API. Otherwise, it's not that obvious if we didn't break anything. 

I also would love to avoid using snapshots for `Slot` tests as they become very hard to reason after changes introduced. In general they are awesome for UI components, but `Slot` is more container which doesn't care as much about UI but to ensure that all fills are properly rendered.

**Comment by nosolosw** [comment]
@gziolo Tests are now in a separate PR at https://github.com/WordPress/gutenberg/pull/11314 I've rebased this with that PR, so we can merge cleanly after that lands into master.

File: packages/components/src/slot-fill/test/slot.js
Comment (nosolosw) [reply packages/components/src/slot-fill/test/slot.js]: Let's discuss it in https://github.com/WordPress/gutenberg/pull/11314

File: packages/components/src/slot-fill/slot.js
Comment (nosolosw) [reply packages/components/src/slot-fill/slot.js]: 180351cd8c67d5f16b64c13bfcc13814401d1aff

File: packages/components/src/slot-fill/index.js
Comment (nosolosw) [reply packages/components/src/slot-fill/index.js]: Actually, we don't need to export the `Consumer` directly because we're already exporting specialized consumers (`Slot`, and `Fill`). Removed it.

**Review by gziolo — APPROVED** [review]

**Comment by gziolo** [comment]
Nice work, changes in this PR looks great. There are a few e2e tests which ensure that plugins can extend editor, so I think it also covers Slot and FIll logic indirectly.

File: packages/components/src/slot-fill/fill.js
Comment (gziolo) [inline packages/components/src/slot-fill/fill.js]: You can remove `noop` also here for consistency.

Comment (nosolosw) [reply packages/components/src/slot-fill/fill.js]: cb19a4e

**Comment by nosolosw** [comment]
Cool! The base branch is #11314 I'll wait until that's approved and merged, then I'll rebase this with master and will merge.

**Pull request merged**
